src/Shared/Components/PooledByteBufferWritter.cs

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+
+namespace Microsoft.AspNetCore.Components.Infrastructure
+{
+    internal sealed class PooledByteBufferWriter : IBufferWriter<byte>, IDisposable
+    {
+        private byte[] _currentBuffer;
+        private int _index;
+        private bool _owned;
+
+        private const int MinimumBufferSize = 256;
+
+        public PooledByteBufferWriter(int initialCapacity = 0)
+        {
+            _currentBuffer = ArrayPool<byte>.Shared.Rent(initialCapacity);
```

Comment:
This is a bit iffy - in the usage, the lifetime of `PooledByteBufferWriter` is not super crisp.  If I understand correctly, every time we call `Persist` we would rent from the array pool, but `Pause` must be invoked for these buffers to be disposed  / returned. If the latter does not happen, we would have a leak.

In Mvc, we use a custom set of pools that are tied to the lifetime of the current request. At least that way we know we're not leaking anything beyond the scope of what we're comfortable with. Could we do something similar here and tie to this to the circuit?

Reply:
I've updated this to make the ComponentStatePersistenceManager implementation `IDisposable` and clear the rented buffers at that time, which will in turn force the release at the end of the request/circuit.

That said, Pause is only called within PersistStateAsync which is invoked by the framework.

Reply:
I've tightened things up to make sure you can't only call Persist within an OnPersisting callback invocation, this makes sure that a buffer is only rented for a short period of time.